index.rst

```diff
@@ -1,29 +1,375 @@
+
+There are two basic ways to respond to a user hitting a rate limit: delay the request until the rate limit allows it, or reject the request.
+We've chosen to reject the request, since delaying it requires queuing it on the server, which adds load to a potentially already-overloaded server and also may create complex timeout issues that are hard to debug.
+
+The drawback of rejecting the request is that it may produce failures in long-running processes when one of their underlying requests is rate-limited.
+This will place the onus on the user to retry rate-limited requests if needed.
+We may want to add support to PyVO_ for retrying rate-limited requests automatically.
+
+.. _PyVO: https://pyvo.readthedocs.io/en/latest/
+
+Metrics
+=======
+
+Rate limit information will be logged as part of the log message for each authentication request to Gafaelfawr.
+
+We will eventually want more data than that, particularly for rate limiting.
+Ideally, Gafaelfawr should log metrics for how many users are being rate-limited, how many requests were rejected due to rate-limiting (and from how many distinct users), and how many users have reached 50% or 75% of a rate limit.
+We don't yet have a general metrics framework for Gafaelfawr; once one exists, metrics like that will be added.
+
+.. _dos-protection:
+
+Denial of service protection
+============================
 
-Requirements, options, and proposed design for quota management and API rate limiting in the Rubin Science Platform.
+This rate limiting system is intended to fairly share resources among non-malicious users issuing a normal rate of API requests.
+Each request, even if rate-limited, requires processing by NGINX, an auth subrequest to Gafaelfawr, and processing by Gafaelfawr, including at least two Redis reads, one write, and often an LDAP lookup.
+This means NGINX and Gafaelfawr could still be overloaded by higher quantities of traffic, such as runaway processes in tight loops or an intentional denial of service attack.
 
-Add content here
-================
+Fully defending against denial of service attacks is outside the scope of the Rubin Science Platform and not something we can reasonably expect to do.
+But we can apply sanity limits on requests at the NGINX level to protect against being overwhelmed by accidents external to the cluster.
 
-Add content here.
-See the `reStructuredText Style Guide <https://developer.lsst.io/restructuredtext/style.html>`__ to learn how to create sections, links, images, tables, equations, and more.
+This can be done with `ingress-nginx annotations <https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#rate-limiting>`__, normally managed via ``GafaelfawrIngress``.
+This rate limiting can only be done by IP address, not by user.
+The NGINX rate limit should be higher than the quota of any given user, since it will be applied to every user and may apply to multiple users at the same time if they share an outbound IP address.
```

Comment:
Ah, OK, that's my previous objection.  In practice I'm not sure we're going to be able to distinguish between scichris (see #ops-usdf, Nov. 16 2022) and a DoS attack and if scichris is at an institution with a single egress address...well, I guess we'll see what happens.